src/index.js

```diff
@@ -73,10 +73,26 @@ function optimizeAndCompileSchema({schema, profileDocuments, compiler}) {
 }
 
 function prependFragments(source, id) {
+  return recursivelySourceFragmentFiles(source, id).then((fragments) => {
+    return source.concat(fragments.map((fragment) => fragment.body).join('\n'));
+  });
+}
+
+function recursivelySourceFragmentFiles(source, id) {
   const fragmentFiles = fragmentFilesForDocument(id, source);
 
+  if (fragmentFiles.length === 0) {
+    return Promise.resolve([]);
+  }
+
   return readFiles(fragmentFiles).then((fragments) => {
-    return source.concat(fragments.map((fragment) => fragment.body).join('\n'));
+    return Promise.all(fragments.map((fragment) => {
+      return recursivelySourceFragmentFiles(fragment.body, fragment.path);
+    }).concat(Promise.resolve(fragments)));
+  }).then((fragmentLists) => {
+    return fragmentLists.reduce((acc, fragments) => {
+      return acc.concat(fragments);
+    }, []);
```

Comment:
@swalkinshaw this is the primary change. If you can think of a better way, please let me know. This is asynchronously recursive, so it's kinda hard to follow.